pkg/api/grpc/grpc.go

```diff
@@ -766,31 +768,32 @@ func (a *api) SaveState(ctx context.Context, in *runtimev1pb.SaveStateRequest) (
 	diag.DefaultComponentMonitoring.StateInvoked(ctx, in.GetStoreName(), diag.Set, err == nil, elapsed)
 
 	if err != nil {
-		err = a.stateErrorResponse(err, messages.ErrStateSave, in.GetStoreName(), err.Error())
+		err, grpcCode := a.stateErrorResponse(err, messages.ErrStateSave, in.GetStoreName(), err.Error())
+		err = apierrors.Basic(grpcCode, http.StatusInternalServerError, errorcodes.StateSave, fmt.Sprintf(messages.ErrStateSave, in.GetStoreName(), err.Error()))
 		a.logger.Debug(err)
 		return empty, err
 	}
 	return empty, nil
 }
 
 // stateErrorResponse takes a state store error, format and args and returns a status code encoded gRPC error.
-func (a *api) stateErrorResponse(err error, format string, args ...interface{}) error {
+func (a *api) stateErrorResponse(err error, format string, args ...interface{}) (error, codes.Code) {
 	var etagErr *state.ETagError
 	if errors.As(err, &etagErr) {
 		switch etagErr.Kind() {
 		case state.ETagMismatch:
-			return status.Errorf(codes.Aborted, format, args...)
+			return status.Errorf(codes.Aborted, format, args...), codes.Aborted
 		case state.ETagInvalid:
-			return status.Errorf(codes.InvalidArgument, format, args...)
+			return status.Errorf(codes.InvalidArgument, format, args...), codes.InvalidArgument
```

Comment:
Wouldn't it be simpler if we updated this method to return the Basic() error instead?